server/src/main/java/io/druid/server/security/PreResponseAuthorizationCheckFilter.java

```diff
@@ -132,12 +110,41 @@ public void destroy()
 
   }
 
-  private void handleAuthorizationCheckError(String errorMsg, ServletResponse servletResponse)
+  private void handleUnauthenticatedRequest(
+      final HttpServletRequest request,
+      final HttpServletResponse response
+  ) throws IOException
   {
-    log.error(errorMsg);
+    // Since this is the last filter in the chain, some previous authentication filter
+    // should have placed an authentication result in the request.
+    // If not, send an authentication challenge.
+    Set<String> supportedAuthSchemes = Sets.newHashSet();
+    for (Authenticator authenticator : authenticators) {
+      String challengeHeader = authenticator.getAuthChallengeHeader();
+      if (challengeHeader != null) {
+        supportedAuthSchemes.add(challengeHeader);
+      }
+    }
+    for (String authScheme : supportedAuthSchemes) {
+      response.addHeader("WWW-Authenticate", authScheme);
+    }
+    QueryInterruptedException unauthorizedError = new QueryInterruptedException(
+        QueryInterruptedException.UNAUTHORIZED,
+        null,
+        null,
+        DruidNode.getDefaultHost()
+    );
+    unauthorizedError.setStackTrace(new StackTraceElement[0]);
+    OutputStream out = response.getOutputStream();
+    sendJsonError(response, Response.SC_UNAUTHORIZED, jsonMapper.writeValueAsString(unauthorizedError), out);
+    out.close();
+    return;
+  }
 
+  private void handleAuthorizationCheckError(String errorMsg, ServletResponse servletResponse)
+  {
     // Send out an alert so there's a centralized collection point for seeing errors of this nature
-    log.makeAlert(errorMsg);
+    log.makeAlert(errorMsg).emit();
```

Comment:
This should be split into something like `log.makeAlert(errorMsg).addData("uri",  uri).addData("method", method).emit()`. The reason is that having a consistent message helps identify these after the alerts have been collected. They're easy to search for and then the "uri" field can be inspected to discover which endpoint is bad. "method" is important too so we can differentiate get/post/delete.